cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java

```diff
@@ -283,51 +236,62 @@ public void setStop(final Id.Program id, final String pid, final long endTime, f
   public void setStop(final Id.Program id, final String pid, final long endTime, final ProgramRunStatus runStatus,
                       final BasicThrowable failureCause) {
     Preconditions.checkArgument(runStatus != null, "Run state of program run should be defined");
-    appsTx.get().executeUnchecked(new TransactionExecutor.Function<AppMetadataStore, Void>() {
+    txExecute(transactional, new TxRunnable() {
       @Override
-      public Void apply(AppMetadataStore mds) throws Exception {
-        mds.recordProgramStop(id, pid, endTime, runStatus, failureCause);
-        return null;
-      }
-    }, apps.get());
+      public void run(DatasetContext context) throws Exception {
+        AppMetadataStore metaStore = getAppMetadataStore(context);
+        metaStore.recordProgramStop(id, pid, endTime, runStatus, failureCause);
+
+        // This block has been added so that completed workflow runs can be logged to the workflow dataset
+        if (id.getType() == ProgramType.WORKFLOW && runStatus == ProgramRunStatus.COMPLETED) {
+          RunRecordMeta runRecordMeta = metaStore.getRun(id, pid);
+          if (runRecordMeta == null) {
```

Comment:
did you add this here to avoid the extra lookup in recordCompletedWorkflow()? In that case, you can remove it there, right?